### PR TITLE
Add support for Redis Sentinel

### PR DESCRIPTION
**Pull request by rakyi**
I'm opening this PR to get feedback and start a discussion about the design of Redis Sentinel support, see #80. I tried to reuse maximum of the existing functionality, however the way Sentinel is implemented in aioredis is a poor fit for the existing architecture.

When getting a reference to Redis client using `sentinel.master_for('mymaster')` aioredis returns a [`ManagedPool`](https://github.com/aio-libs/aioredis/blob/fc17ab03ea9ba63175664fc06735146850ed1789/aioredis/sentinel/pool.py#L379) instance which should automatically reconnect to a new master after a failover. This is nice in theory, but we can't specify the loop to use and thus are unable to easily reuse our `ConnectionPool`.

When testing the naive approach in my first commit by swapping the layer in our project, which uses [DjangoChannelsGraphqlWs](https://github.com/datadvance/DjangoChannelsGraphqlWs), I get the following exception and I don't know who's to blame:

```
An error occurred while resolving field GraphQLSubscription.notification
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/graphql/execution/executor.py", line 447, in resolve_or_error
    return executor.execute(resolve_fn, source, info, **args)
  File "/usr/local/lib/python3.7/site-packages/graphql/execution/executors/asyncio.py", line 69, in execute
    result = fn(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/channels_graphql_ws/graphql_ws.py", line 466, in _subscribe
    return register(groups, publish_callback, unsubscribed_callback)
  File "/usr/local/lib/python3.7/site-packages/channels_graphql_ws/graphql_ws.py", line 889, in register
    group, self.channel_name
  File "/usr/local/lib/python3.7/site-packages/asgiref/sync.py", line 64, in __call__
    return call_result.result()
  File "/usr/local/lib/python3.7/concurrent/futures/_base.py", line 425, in result
    return self.__get_result()
  File "/usr/local/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.7/site-packages/asgiref/sync.py", line 78, in main_wrap
    result = await self.awaitable(*args, **kwargs)
  File "/app/channels_redis/channels_redis/core.py", line 544, in group_add
    await connection.zadd(group_key, time.time(), channel)
RuntimeError: Task <Task pending coro=<AsyncToSync.main_wrap() running at /usr/local/lib/python3.7/site-packages/asgiref/sync.py:78> cb=[_run_until_complete_cb() at /usr/local/lib/python3.7/asyncio/base_events.py:158]> got Future <Future pending> attached to a different loop
```

If we'd add the Sentinel support we should also create extensive tests for failovers etc, but I've no idea if that's possible without running the actual Sentinel and Redis processes.

### CODE REVIEW

**Comment by chrismeyersfsu** [comment]
Channels suffered from this same problem and was fixed in channels 2.0.3 via the following commit https://github.com/django/channels_redis/commit/f5e4799e11f472cc267598e9f78099a160f81550 Since this commit, `channels_redis` has grown in complexity so it's a bit hard to follow how this re-alignment exists in the most recent code; but it does still exist.

I _think_ it has to do with `channels_redis` supporting `aync_to_sync` + threads.

**Comment by qeternity** [comment]
What's the latest on this? It doesn't appear that any progress has been made. This seems to be a large blocking item to running channels in production as without Sentinel/Cluster support there's no HA.

**Comment by severindellsperger** [comment]
Any updates here?
It seems that many people like to have this feature.

**Comment by rakyi** [comment]
I didn't pursue this further since I didn't get a reply from maintainers. We ended up using https://github.com/CJWorkbench/channels_rabbitmq to get HA instead.

**Comment by severindellsperger** [comment]
> I didn't pursue this further since I didn't get a reply from maintainers. We ended up using https://github.com/CJWorkbench/channels_rabbitmq to get HA instead.

Sad to hear this...
Unfortunately, we have a Redis cluster in our environment and can't switch to RabbitMQ, but thanks for your suggestion.

Hopefully, a maintainer will come back to this thread.

**Comment by carltongibson** [comment]
> What´s the latest...

Well, it's marked draft for a start, so is it finished? What tests would we add? What about some documentation? 
Really anyone is welcome to take this on and push it forwards. 

`channels_redis` is one file, so you're very welcome to use this in your project, fix issues, report success and champion the feature. If you get involved you'll likely find it moves forward.

**Comment by severindellsperger** [comment]
> I didn't pursue this further since I didn't get a reply from maintainers. We ended up using https://github.com/CJWorkbench/channels_rabbitmq to get HA instead.

It really seems the mentioned library has great support for HA! Because I have no time to do the changes in this library, I will also stick with RabbitMQ-library and maybe will come back later.
Anyway, thanks again for the hint.

**Comment by rakyi** [comment]
Sentinel support was added in #257.
